File: src/limiters/tokenBucket.lua
-- Token bucket, evaluated atomically inside Redis so concurrent requests
-- from the same client can never race each other (no separate GET then SET).
--
-- KEYS[1] = bucket key, e.g. "ratelimit:tb:user:42"
-- ARGV[1] = capacity (max tokens)
-- ARGV[2] = refillRatePerSec
-- ARGV[3] = nowMs
-- ARGV[4] = cost
--
-- Returns: { allowed(0/1), tokensRemaining, retryAfterMs }
--
-- This mirrors src/utils/algorithms.js#tokenBucketCalc exactly — that file
-- is unit tested; keep the two in sync if you change the math.

local key = KEYS[1]
local capacity = tonumber(ARGV[1])
local refillRatePerSec = tonumber(ARGV[2])
local now = tonumber(ARGV[3])
local cost = tonumber(ARGV[4])

if cost > capacity then
  return { 0, 0, -1 }
end

local data = redis.call('HMGET', key, 'tokens', 'ts')
local tokens = tonumber(data[1])
local lastRefill = tonumber(data[2])

if tokens == nil then
  tokens = capacity
  lastRefill = now
end

local elapsedSec = math.max(0, now - lastRefill) / 1000
local refilled = math.min(capacity, tokens + elapsedSec * refillRatePerSec)

local allowed = 0
local retryAfterMs = 0

if refilled >= cost then
  allowed = 1
  refilled = refilled - cost
else
  local deficit = cost - refilled
  retryAfterMs = math.ceil((deficit / refillRatePerSec) * 1000)
end

redis.call('HMSET', key, 'tokens', refilled, 'ts', now)
-- Auto-expire an idle bucket once it would have fully refilled anyway,
-- plus a small buffer, so we don't keep state around forever for clients
-- who never come back.
local ttlSec = math.ceil(capacity / refillRatePerSec) + 60
redis.call('EXPIRE', key, ttlSec)

return { allowed, math.floor(refilled), retryAfterMs }

File: src/limiters/tokenBucket.js
'use strict';

/**
 * Variable-cost limiter backed by the token-bucket Lua script. Use for
 * anything priced per-unit rather than per-request — the canonical case
 * here is LLM token usage.
 */
class TokenBucketLimiter {
  /**
   * @param {import('ioredis').Redis} redis - client with the `tokenBucket` command defined
   * @param {Object} opts
   * @param {number} opts.capacity - max tokens the bucket can hold
   * @param {number} opts.refillRatePerSec - tokens added per second
   * @param {string} [opts.keyPrefix]
   */
  constructor(redis, { capacity, refillRatePerSec, keyPrefix = 'ratelimit:tb' }) {
    this.redis = redis;
    this.capacity = capacity;
    this.refillRatePerSec = refillRatePerSec;
    this.keyPrefix = keyPrefix;
  }

  /**
   * @param {string} identifier - e.g. user id, API key, IP
   * @param {number} cost - tokens this request consumes
   * @returns {Promise<{allowed: boolean, remaining: number, retryAfterMs: number, limit: number}>}
   */
  async consume(identifier, cost = 1) {
    const key = `${this.keyPrefix}:${identifier}`;
    const now = Date.now();
    const [allowed, remaining, retryAfterMs] = await this.redis.tokenBucket(
      key,
      this.capacity,
      this.refillRatePerSec,
      now,
      cost
    );
    return {
      allowed: allowed === 1,
      remaining,
      retryAfterMs: retryAfterMs < 0 ? Infinity : retryAfterMs,
      limit: this.capacity,
    };
  }
}

module.exports = TokenBucketLimiter;
